Guled/finish creating add comic form

Somnibyte opened the pull request:
The following update includes:

* Complete *Add Comic* form based on [this design](https://app.zeplin.io/project/59c169df713d751d2245ccd0/screen/59cbe4c0966b0e68ac1725a4).
* Functional drop down picker views for the *release date*, *genre* and *condition* fields. (Scroll view does not react to the height of the pickers change yet.)

~Note that the navbar is not included in this update.~

Commit:
Add AddComicTableViewController and AddCover Cell.

Commit:
Add AddComicTableViewController and AddComicTableViewCell prototype.

Add AddComicViewController with tableview and custom cells to emulate form fields.

Adjust Select a Date button image edge inset and position. Edit Spacing between RELEASE and Select a Date button.

Commit:
Utilize stackviews rather than tableviews to create form.

Add NavigationBar. Customize NavigationBar. Add views under pickerviews. Add border radius to cover photo button.

Remove images that are no longer being used.

Somnibyte commented:
Made a couple of changes to make this PR more complete. 

* Navbar is now included.
* Genre and Condition fields now have functional drop down functionality when tapping the *Select ...* button.
* Overall design has been polished to emulate designs found on Zeplin.

mayasaxena commented on lucien-ios.xcodeproj/xcuserdata/guled.ahmed.xcuserdatad/xcschemes/xcschememanagement.plist:
```diff
@@ -7,7 +7,7 @@
 		<dict>
 			<key>orderHint</key>
-			<integer>3</integer>
+			<integer>6</integer>
```
This change should be reverted - if you don't know how I can show you

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,208 @@
+
+    // MARK: - Constants
+
+    private let conditions = ["Near Mint",
```
The formatting on the array is a little off - each element should be on its own line, indented by one tab from the start of the first line e.g.
```
private let conditions = [
    "Near Mint",
    "Very Fine",
    "Fine",
    "Very Good",
    "Good",
    "Fair",
    "Poor"
]
```

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,208 @@
+                             "Poor"
+    ]
+
+    private let genres = ["Action/Adventure",
```
Same as above, but let's chat about how/where we're storing this. This type of data, especially since you'll be referring to it later, is better suited to an enum rather than just a list of strings

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,208 @@
+
+    private func configureViewController() {
+        // Cover Photo Cell
+        coverPhotoLabel.font = UIFont(name: LucienTheme.Fonts.Muli.light.rawValue, size: 12)
```
We'll discuss this further, but usually the idea with a Theme file is that it provides you the font itself, not just the name, so you don't have to create it on the fly

Somnibyte replied:
I couldn't make an enum with UIFonts, but I think an alternative is to have a struct that provides the UIFonts directly. Would this be a better idea?

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,208 @@
+    private func configureViewController() {
+        // Cover Photo Cell
+        coverPhotoLabel.font = UIFont(name: LucienTheme.Fonts.Muli.light.rawValue, size: 12)
+        coverPhotoLabel.textColor = LucienTheme.coolGrey
```
Very nice use of Theme!

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,208 @@
+        coverPhotoLabel.font = UIFont(name: LucienTheme.Fonts.Muli.light.rawValue, size: 12)
+        coverPhotoLabel.textColor = LucienTheme.coolGrey
+
+        coverPhotoButton.titleLabel?.lineBreakMode = .byWordWrapping
```
A lot of these things can be set in the nib itself rather than programmatically. My rule of thumb is that if we're bothering to use nibs, we should take advantage of them as much as we can and only set things programmatically that can't be handled in the nib

Somnibyte replied:
If you are referring to the fonts as well I'd like to note that for some reason custom fonts, despite creating a directory for the fonts and editing the info.plist, won't show up on the Attributes Inspector. Have you had this issue before? I haven't been able to fix this yet.

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,208 @@
+        coverPhotoButton.layer.borderColor = UIColor.gray.cgColor
+
+        // Comic Title
+        cconfigureLabelwith(label: comicTitleLabel, fontName: LucienTheme.Fonts.Muli.light.rawValue, size: 12, color: LucienTheme.coolGrey)
```
Really good thought with this helper function, but let's refactor it a little. Also watch out for typos

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,208 @@
+    // MARK: - IBOutlet Methods
+
+    @IBAction func selectADateButtonTapped(_ sender: UIButton) {
+        UIView.animate(withDuration: 0.3, animations: { [weak self] in
```
Whenever possible, try to use trailing closures. Also, there's no risk for a retain cycle in `UIView.animate` so you don't need to capture self weakly e.g.
```
UIView.animate(withDuration: 0.3) {
    self.datePicker.isHidden = false
}
```

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,208 @@
+    }
+
+    @IBAction func selectGenreButtonTapped(_ sender: UIButton) {
+        UIView.animate(withDuration: 0.3, animations: { [weak self] in
```
Same as above

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,208 @@
+    }
+
+    @IBAction func selectConditionButtonTapped(_ sender: UIButton) {
+        UIView.animate(withDuration: 0.3, animations: { [weak self] in
```
Same as above

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,208 @@
+    }
+
+    // MARK: - UIPickerViewDataSource
+    func numberOfComponents(in pickerView: UIPickerView) -> Int {
```
There should be a newline after the `MARK`

mayasaxena commented on lucien-ios/LucienTheme.swift:
```diff
@@ -0,0 +1,55 @@
+     The createUIColor achieves this by dividing each RGB value by 255.
+     */
+    convenience init(redVal: CGFloat, greenVal: CGFloat, blueVal: CGFloat, alphaVal: CGFloat) {
+        self.init(red: CGFloat(redVal)/255, green: CGFloat(greenVal)/255, blue: CGFloat(blueVal)/255, alpha: alphaVal)
```
Same comment as I gave for Gracie, there should be spaces around mathematical operators

mayasaxena commented on lucien-ios/StartMyCollection/StartMyCollectionViewController.swift:
```diff
@@ -10,5 +10,32 @@ import UIKit
+        present(addComicNavigationController, animated: true, completion: nil)
+    }
+
+    @objc func backButtonTapped(){
```
There should be a space after the `()`

mayasaxena commented on lucien-ios/ViewModels/AddCoverCellViewModel.swift:
```diff
@@ -0,0 +1,26 @@
+import QuartzCore
+
+struct AddCoverCellViewModel {
+    /// Configures fonts and colors for a AddCoverTableViewCell. 
```
Commented-out code should not be in a PR

mayasaxena commented on lucien-ios/ViewModels/ComicPickerViewCellViewModel.swift:
```diff
@@ -0,0 +1,73 @@
+import Foundation
+import UIKit
+
+//struct ComicPickerViewCellViewModel {
```
Commented-out code should not be in a PR

mayasaxena commented on lucien-ios/ViewModels/ComicReleaseDateCellViewModel.swift:
```diff
@@ -0,0 +1,25 @@
+import UIKit
+
+struct ComicReleaseDateCellViewModel {
+    /// Configures fonts, colors and border for a ComicReleaseDateCell
```
Commented-out code should not be in a PR

mayasaxena commented on lucien-ios/ViewModels/ComicTextFieldCellViewModel.swift:
```diff
@@ -0,0 +1,58 @@
+import UIKit
+
+struct ComicTextFieldCellViewModel {
+//    private let fieldTitleArray = [1:"COMIC TITLE", 2:"STORY TITLE", 3:"VOLUME", 4:"ISSUE", 5:"PUBLISHER", 7:"CONDITION", 8: "GENRE"]
```
Commented-out code should not be in a PR

mayasaxena commented on lucien-ios/StartMyCollection/StartMyCollectionViewController.swift:
```diff
@@ -10,5 +10,32 @@ import UIKit
+        let addComicNavigationController = UINavigationController(rootViewController: addComicViewController)
+
+        // NavigationBar Background
+        addComicNavigationController.navigationBar.backgroundColor = LucienTheme.white
```
The majority of this can be handled in `AddComicViewController` itself. Unless it's something that only `StartMyCollectionViewController` knows about, it shouldn't be handled here

Commit:
Add getFont function to LucientTheme fonts. Convert genre and conditions array to enums. Replace cover button image. Remove unused files and directories. Add finishButton functionality. Make scrollview react to button taps.

Add textfield bottom border highlight effect when selected.

Disable AddComicViewController finishbutton.

Add older version of add cover button. Rename configure methods to better reflect what is being changed.

Add new button image. Adjust add cover label and add cover button constraints.

Make Finish button responsive when comicTitleTextField has been edited.

Add warning label and ability to warn user when comicTitleTextField is incomplete.

Disable finishButton when comicTitleTextField is incomplete.

Remove unused viewmodel directory

Remove Protocol directory. Rename Add New Comic Book directory.

Prevent use from being able to erase comic title and tap finish button.

Remove force unwrap of UIFonts in LucienTheme.

Scroll towards bottom of view when date selected.

Scroll to bottom of view when genre and condition are selected.

Remove unnecessary space.

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,194 @@
+        coverPhotoButton.titleEdgeInsets.top = 25
+
+        // Comic Title
+        configureLabel(label: comicTitleLabel, font: LucienTheme.Fonts.Muli.light.getFont(size: 12), color: LucienTheme.coolGrey)
```
I would suggest splitting each section with a comment in front of it into its own function

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,194 @@
+    // MARK: - IBOutlet Methods
+
+    @IBAction func selectADateButtonTapped(_ sender: UIButton) {
+        UIView.animate(withDuration: 0.3, animations: {
```
Need to use trailing closure syntax here e.g.
```
UIView.animate(withDuration: 0.3) {
    self.datePicker.isHidden = false
}

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,194 @@
+
+        // NavigationBar 'Finish' Button
+        let finishButton = UIBarButtonItem(title: "Finish", style: .plain, target: self, action: nil)
+        finishButton.setTitleTextAttributes([NSAttributedStringKey.font: LucienTheme.Fonts.Muli.semiBold.getFont(size: 17)], for: .normal)
```
Having this `getFont` method is fine, but I'd suggest helper methods in theme for the most common fonts so you can just call `LucienTheme.muliSemiBold(size: 17)` which is a little less verbose.

Somnibyte commented:
The form is now complete. The required field (comic title) is now functional and can detect whether or not it has been edited which, in turn, triggers the *finish* button at the top right corner of the screen to be enabled.

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.xib:
```diff
@@ -0,0 +1,372 @@
+                                                <action selector="selectADateButtonTapped:" destination="-1" eventType="touchUpInside" id="4r4-bM-0jR"/>
+                                            </connections>
+                                        </button>
+                                        <datePicker hidden="YES" contentMode="scaleToFill" contentHorizontalAlignment="center" contentVerticalAlignment="center" datePickerMode="dateAndTime" minuteInterval="1" translatesAutoresizingMaskIntoConstraints="NO" id="Ayn-n0-Cph">
```
Your datePicker probably shouldn't include time as well. That might be useful for the due date but I don't think release time is something that's known

mayasaxena commented on lucien-ios.xcodeproj/project.pbxproj:
```diff
@@ -96,6 +138,21 @@
 /* End PBXFrameworksBuildPhase section */
 
 /* Begin PBXGroup section */
+		1C07CBFE1F7BEC8D007C66D0 /* ViewModels */ = {
```
Looks like you deleted this folder - make sure to delete the group in Xcode too

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,224 @@
+
+        // NavigationBar 'Finish' Button
+        finishButton = UIBarButtonItem(title: "Finish", style: .plain, target: self, action: nil)
+        finishButton.setTitleTextAttributes([NSAttributedStringKey.font: LucienTheme.Fonts.Muli.semiBold.getFont(size: 17)], for: .normal)
```
Having this `getFont` method is fine, but I'd suggest helper methods in theme for the most common fonts so you can just call `LucienTheme.muliSemiBold(size: 17)` which is a little less verbose.

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,224 @@
+
+        // NavigationBar Background
+
+        navigationController.navigationBar.isTranslucent = true
```
Every section that's more than one line and headed by a comment should be extracted into a function.

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,224 @@
+        button.titleEdgeInsets.top = 25
+    }
+
+    private func createPlaceHolderAttributedString(textFieldPlaceholderText: String) -> NSAttributedString {
```
The parameter name is a little verbose, you can probably just use `placeholder`

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,224 @@
+
+    private func createPlaceHolderAttributedString(textFieldPlaceholderText: String) -> NSAttributedString {
+        let placeholderAttributes = [
+        NSAttributedStringKey.foregroundColor: LucienTheme.silver,
```
These dictionary elements should each by indented one level from the `let`

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,224 @@
+    }
+
+    @IBAction func selectGenreButtonTapped(_ sender: UIButton) {
+        UIView.animate(withDuration: 0.3, animations: {
```
Same as above

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,224 @@
+    }
+
+    @IBAction func selectConditionButtonTapped(_ sender: UIButton) {
+        UIView.animate(withDuration: 0.3, animations: {
```
Same as above

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,224 @@
+        bottomBorder.backgroundColor = LucienTheme.dark.cgColor
+
+        guard let comicTitleText = comicTitleTextField.text else {
+            return
```
This `return` can be on the same line as the `guard`. You only need to put it on a new line if it's doing something other than returning

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,224 @@
+            comicTitleTextFieldBottomBorder.backgroundColor = LucienTheme.textFieldBottomBorderWarning.cgColor
+            finishButton.isEnabled = false
+            finishButton.tintColor = LucienTheme.finishButtonGrey
+        }else if textField == comicTitleTextField {
```
Need a space before `else`

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,224 @@
+    }
+
+    func textFieldDidEndEditing(_ textField: UITextField) {
+        let bottomBorder = textField.layer.sublayers![0] as CALayer
```
Don't force unwrap here

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,224 @@
+
+        if textField != comicTitleTextField && comicTitleText.isEmpty {
+            comicTitleWarningLabel.isHidden = false
+            let comicTitleTextFieldBottomBorder = comicTitleTextField.layer.sublayers![0] as CALayer
```
Don't force unwrap here

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,224 @@
+    // MARK: - UITextFieldDelegate
+
+    func textFieldDidBeginEditing(_ textField: UITextField) {
+        let bottomBorder = textField.layer.sublayers![0] as CALayer
```
Don't force unwrap here

mayasaxena commented on lucien-ios/LucienTheme.swift:
```diff
@@ -0,0 +1,96 @@
+            func getFont(size: CGFloat) -> UIFont {
+                switch self {
+                case .black :
+                    return UIFont(name: rawValue, size: size)!
```
Don't force unwrap the font - return an optional

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.xib:
```diff
@@ -0,0 +1,408 @@
+                <scrollView clipsSubviews="YES" multipleTouchEnabled="YES" contentMode="scaleToFill" translatesAutoresizingMaskIntoConstraints="NO" id="8pQ-9n-KZP">
+                    <rect key="frame" x="0.0" y="0.0" width="375" height="1677"/>
+                    <subviews>
+                        <label opaque="NO" userInteractionEnabled="NO" contentMode="left" horizontalHuggingPriority="251" verticalHuggingPriority="251" text="COVER PHOTO" textAlignment="natural" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" translatesAutoresizingMaskIntoConstraints="NO" id="ZeW-7M-5vu">
```
You have constraint conflicts in this file. They need to be fixed before putting up a PR

mayasaxena commented on lucien-ios/Add New Comic Book/AddComicViewController.swift:
```diff
@@ -0,0 +1,224 @@
+        button.titleLabel?.lineBreakMode = .byWordWrapping
+        button.titleLabel?.textAlignment = .center
+        button.setTitle("Take \n Cover \n Photo", for: .normal)
+        button.titleLabel?.font = LucienTheme.Fonts.Muli.bold.getFont(size: 13)
```
Your fonts aren't showing up in Interface Builder because they're not installed on your computer itself. If you need help with that, let me know

mayasaxena commented on lucien-ios/AddNewComicBook/AddComicViewController.swift:
```diff
@@ -0,0 +1,233 @@
+        }
+
+        // NavigationBar Background
+        navigationController.navigationBar.isTranslucent = true
```
Each chunk of lines headed with a comment should be extracted into a function

mayasaxena commented on lucien-ios/AddNewComicBook/AddComicViewController.swift:
```diff
@@ -0,0 +1,233 @@
+
+    private func configureNavigationController() {
+        guard let navigationController = navigationController else {
+            return
```
This can be on the same line as the `guard`

mayasaxena commented on lucien-ios/AddNewComicBook/AddComicViewController.swift:
```diff
@@ -0,0 +1,233 @@
+        button.tintColor = LucienTheme.dark
+    }
+
+    private func configureAddCoverUIButton(button: UIButton) {
```
Unless you're reusing this function with another cover photo button, you don't need to pass in the button, you can just use `coverPhotoButton` directly

mayasaxena commented on lucien-ios/AddNewComicBook/AddComicViewController.swift:
```diff
@@ -0,0 +1,233 @@
+    private func configureAddCoverUIButton(button: UIButton) {
+        button.titleLabel?.lineBreakMode = .byWordWrapping
+        button.titleLabel?.textAlignment = .center
+        button.setTitle("Take \n Cover \n Photo", for: .normal)
```
You can set this and everything after it in Interface Builder directly

mayasaxena commented on lucien-ios/AddNewComicBook/AddComicViewController.swift:
```diff
@@ -0,0 +1,233 @@
+    // MARK: - IBOutlet Methods
+
+    @IBAction func selectADateButtonTapped(_ sender: UIButton) {
+        UIView.animate(withDuration: 0.3, animations: {
```
This should be formatted differently, when you have multiple closures, each argument should be on its own line indented once from the top line e.g.
```
UIView.animate(
    withDuration: 0.3, 
    animations: {
        self.datePicker.isHidden = false
    },
    completion: { _ in
        let offset = CGPoint(x: 0, y: self.scrollView.contentSize.height - self.scrollView.bounds.size.height)
        self.scrollView.setContentOffset(offset, animated: true)
    }
)
```

mayasaxena commented on lucien-ios/AddNewComicBook/AddComicViewController.swift:
```diff
@@ -0,0 +1,233 @@
+    }
+
+    @IBAction func selectGenreButtonTapped(_ sender: UIButton) {
+        UIView.animate(withDuration: 0.3, animations: {
```
Same as above

mayasaxena commented on lucien-ios/AddNewComicBook/AddComicViewController.swift:
```diff
@@ -0,0 +1,233 @@
+    }
+
+    @IBAction func selectConditionButtonTapped(_ sender: UIButton) {
+        UIView.animate(withDuration: 0.3, animations: {
```
Same as above

mayasaxena commented on lucien-ios/AddNewComicBook/AddComicViewController.swift:
```diff
@@ -0,0 +1,233 @@
+        bottomBorder.backgroundColor = LucienTheme.dark.cgColor
+
+        guard let comicTitleText = comicTitleTextField.text else {
+            return
```
This can be on the same line as the `guard`

mayasaxena commented on lucien-ios/AddNewComicBook/AddComicViewController.swift:
```diff
@@ -0,0 +1,233 @@
+    // MARK: - UITextFieldDelegate
+
+    func textFieldDidBeginEditing(_ textField: UITextField) {
+        let bottomBorder = textField.layer.sublayers![0] as CALayer
```
Don't force unwrap

mayasaxena commented on lucien-ios/AddNewComicBook/AddComicViewController.swift:
```diff
@@ -0,0 +1,233 @@
+
+        if textField != comicTitleTextField && comicTitleText.isEmpty {
+            comicTitleWarningLabel.isHidden = false
+            let comicTitleTextFieldBottomBorder = comicTitleTextField.layer.sublayers![0] as CALayer
```
Don't force unwrap

mayasaxena commented on lucien-ios/AddNewComicBook/AddComicViewController.swift:
```diff
@@ -0,0 +1,233 @@
+    }
+
+    func textFieldDidEndEditing(_ textField: UITextField) {
+        let bottomBorder = textField.layer.sublayers![0] as CALayer
```
Don't force unwrap

mayasaxena commented on lucien-ios/AddNewComicBook/AddComicViewController.swift:
```diff
@@ -0,0 +1,233 @@
+import QuartzCore
+
+final class AddComicViewController: UIViewController, UIPickerViewDelegate, UIPickerViewDataSource, UITextFieldDelegate {
+    // MARK: - Private IBOutlets
```
Need a newline before the mark

mayasaxena commented on lucien-ios/AddNewComicBook/AddComicViewController.swift:
```diff
@@ -0,0 +1,233 @@
+    @IBOutlet private weak var comicTitleWarningLabel: UILabel!
+
+    // MARK: - Private Variables
+    private var didCompleteForm = false
```
Need a newline after the mark

mayasaxena commented on lucien-ios/AddNewComicBook/AddComicViewController.swift:
```diff
@@ -0,0 +1,233 @@
+
+        // NavigationBar 'Finish' Button
+        finishButton = UIBarButtonItem(title: "Finish", style: .plain, target: self, action: nil)
+        finishButton.setTitleTextAttributes([NSAttributedStringKey.font: LucienTheme.Fonts.Muli.semiBold.getFont(size: 17) ?? UIFont()], for: .normal)
```
`getFont` is fine, it gets verbose very quickly so you should have accessor methods to make it easier that can be called like this: `LucienTheme.muliSemiBold(size: 17)`. Better yet, in theme create a static constant named after what the font is for (in this case `finishButtonFont`) which you can access like this `LucienTheme.finishButtonFont`.

mayasaxena commented on lucien-ios/AddNewComicBook/AddComicViewController.swift:
```diff
@@ -0,0 +1,233 @@
+        button.titleEdgeInsets.top = 25
+    }
+
+    private func createPlaceHolderAttributedString(textFieldPlaceholderText: String) -> NSAttributedString {
```
Rename external parameter to `placeholder`

Commit:
Correct syntax. Cleanup viewdidload method. Customize objects in xib.

Fix guard statements.

Create methods to customize navigationcontroller.

Use static methods to represent fonts.

Cover Photo Button Label fully designed in xib.

Change parameter name of createPlaceHolderAttributeString to just placeholder.

Change parameter names for calls made to createPlaceHolderAttributedString.

Fix UIView animate closure format.

Remove force unwrap.
